### PR TITLE
Dc fix add to cart

### LINKED ISSUE #239: Able to add to cart when quant not available
There is only one item available for this SKU:

Screen Shot 2020-04-18 at 7 12 11

But I'm being able to add this to cart many times:

Screen Shot 2020-04-18 at 7 13 07

### PR DESCRIPTION
**Pull request by DavidCuentasMar**
fixes: #239
yotube link: https://youtu.be/jAtmSUSX8F4

Comments: 

- I think we need a better way of showing messages to user
- I'm not sure the way I filtered the stockitem is the best option (It has a bunch of code, the logic is the same but depends on null values)

### CODE REVIEW

File: lib/screens/customer_product_detail_page.dart [hunk 1]
```diff
@@ -440,14 +443,47 @@ class _CustomerProductDetailPageState extends State<CustomerProductDetailPage> {
...
+              stockItem?.size == _selectedItemSize);
+        }
+        if (_stockRequest.color == null && _stockRequest.size != null) {
+          productStock = _product.stock.items
+              .firstWhere((stockItem) => stockItem?.size == _selectedItemSize);
+        }
+        if (_stockRequest.color != null && _stockRequest.size == null) {
+          productStock = _product.stock.items.firstWhere(
+              (stockItem) => stockItem?.color == _selectedItemColor);
+        }
```
Comment (dragonfly91) [inline lib/screens/customer_product_detail_page.dart]: I think this snippet of code:

```
StockItem productStock = StockItem();
        if (_stockRequest.color != null && _stockRequest.size != null) {
          productStock = _product.stock.items.firstWhere((stockItem) =>
              stockItem?.color == _selectedItemColor &&
              stockItem?.size == _selectedItemSize);
        }
        if (_stockRequest.color == null && _stockRequest.size != null) {
          productStock = _product.stock.items
              .firstWhere((stockItem) => stockItem?.size == _selectedItemSize);
        }
        if (_stockRequest.color != null && _stockRequest.size == null) {
          productStock = _product.stock.items.firstWhere(
              (stockItem) => stockItem?.color == _selectedItemColor);
        }
```

gives the same output as the `_getProductStockItem` method in the `cart_page.dart`. How about making this method public and reusing it in both these places?

File: lib/screens/customer_product_detail_page.dart [hunk 2]
```diff
@@ -440,14 +443,47 @@ class _CustomerProductDetailPageState extends State<CustomerProductDetailPage> {
...
+        if (customer.cart.items[productIndex].stockRequest.quantity <
+            productStock.quantity) {
+          customer.cart.items[productIndex].stockRequest.quantity++;
```
Comment (dragonfly91) [inline lib/screens/customer_product_detail_page.dart]: I think this comparison

```
if (customer.cart.items[productIndex].stockRequest.quantity <
            productStock.quantity) {
```

along with the '**getting the product stock item**' can be inside the `_canAddToCart` method. This way you can minimize the edits to the rest of your code

**Review by dragonfly91 — CHANGES_REQUESTED** [review]

File: lib/screens/customer_product_detail_page.dart (hunk 2 above)
Comment (DavidCuentasMar) [reply lib/screens/customer_product_detail_page.dart]: This change implies to make _canAddToCart async so we can use customer instance and the return of this function would be future<bool>. Do you see this okay?

Comment (dragonfly91) [reply lib/screens/customer_product_detail_page.dart]: Why do you need to make it async in order to pass the customer instance?

Comment (DavidCuentasMar) [reply lib/screens/customer_product_detail_page.dart]: _updateCartModel still doing some checks if the product quantity is available or allowed, I thought that " 'getting the product stock item' can be inside the _canAddToCart method
 mean to move those validations inside _canAddToCart. Is that correct?

Comment (dragonfly91) [reply lib/screens/customer_product_detail_page.dart]: Yes, I suggest something like this:

```
bool _canAddToCart(Product _product) {
    if (_product.stock.stockType == StockType.sizeAndColor) {
      return _selectedItemColor != null && _selectedItemSize != null;
    } else if (_product.stock.stockType == StockType.sizeOnly) {
      return _selectedItemSize != null;
    } else if (_product.stock.stockType == StockType.colorOnly) {
      return _selectedItemColor != null;
    }

    int productIndex = customer.cart.items.indexWhere((cartItem) =>
        cartItem.productDocumentId == widget.productDocumentId &&
        cartItem.stockRequest?.color == _selectedItemColor &&
        cartItem.stockRequest?.size == _selectedItemSize);

    bool addToCartSuccess = true;
    if (productIndex != -1) {
      StockItem productStock = getRequestedStockItem(_product, _stockRequest);
      if (customer.cart.items[productIndex].stockRequest.quantity >=
          productStock.quantity) {
        addToCartSuccess = false;
      }
    }

    return addToCartSuccess;
  }
```

**Comment by dragonfly91** [comment]
@DavidCuentasMar since we have a release coming up pretty soon and we would like to get this bug fix in there, I'm going to salvage all that I can from the PR, as there seem to be issues with the merge. Will be closing this PR for now. Thanks for your contribution on this issue!

**Pull request closed without merging**